presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java

```diff
@@ -235,6 +235,9 @@ public PlanOptimizers(Metadata metadata, SqlParser sqlParser, FeaturesConfig fea
                 new TransformCorrelatedSingleRowSubqueryToProject(),
                 new CheckSubqueryNodesAreRewritten(),
                 new PredicatePushDown(metadata, sqlParser),
+                new IterativeOptimizer(
+                        stats,
+                        new PickTableLayout(metadata, sqlParser, true).rules()),
```

Comment:
I didn't bother updating the deprecated PickLayout rule too. We only need this change when other IterativeOptimizers are also on anyway.

Reply:
Can you please extract variable and use it in both places.

Reply:
Can you please explain me why do you execute it here and then why do you overwrite couple of optimizers later? Maybe we could run it once.

Reply:
I run it after every predicate push down because if there's a new filter above the tablescan we may be able to choose a different layout.  And we want to make sure we're choosing the best layout we can so we can get good stats for any cost-based rules that run.  Even if right now there isn't anything that uses it in between the two times, cost-based optimizers can go anywhere, so we want the layout to be correct no matter where you put them.

Reply:
> cost-based optimizers can go anywhere

Are you sure we want to go with such approach? There might be more things than picking table layout which may influence stats quality (like `SimplifyExpressions`). I would prefer to prepare for now a single place, where we know that stats should be good, and there put all the cost-based related rules. How does it sound to you?

Moreover we may want to to put a `checkState` to stats estimation code to prevent running them without a layout. Just in case someone wanted to put cost-related rule into a wrong place.

Reply:
What about cost-based optimizations that could enable choosing a different layout.  I guess we could worry about that when we have such an optimization.

I like the idea of a `checkState` in the stats code.